src/Battle.php

```diff
@@ -0,0 +1,211 @@
+                $this->looser = $this->player->isAlive() ? $this->monster : $this->player;
+                break;
+            }
+        }
+        
+        return $count;
+    }
+    
+    /**
+     * Fights exactly 1 round
+     * @param int $firstDamageRound
+     */
+    protected function fightOneRound(int $firstDamageRound)
+    {
+        // playerDamage is the damage done to the player, to the monster.
+        list($playerDamage, $monsterDamage, $playerAttack) = $this->calculateDamage();
+        
+        // Player does damage to the monster
+        if ($firstDamageRound & self::DAMAGEROUND_PLAYER
+            && $this->player->isAlive()
+            && $this->monster->isAlive()
+        ) {
+            if ($monsterDamage < 0) {
+                // The damage done to the monster is negative.
+                // This means that the monster conters the player's attack
+                $this->player->damage(0 - $monsterDamage);
+            } elseif ($monsterDamage > 0) {
+                // The damage done to the monster is positive.
+                // This means that this is a normal attack
+                $this->monster->damage($monsterDamage);
+            } else {
+                // The damage done to the monster is 0.
+                // We interpretate this as a miss.
+            }
+        }
+        
+        // Monster does damage to the player
+        if ($firstDamageRound & self::DAMAGEROUND_MONSTER
+            && $this->player->isAlive()
+            && $this->monster->isAlive()
+        ) {
+            if ($playerDamage > 0) {
+                // The damage done to the player is negative
+                // THis means that the player conters the monster's attack
+                $this->monster->damage(0 - $playerDamage);
+            } elseif($playerDamage > 0) {
+                // The damage done to the player is positive.
+                // This means that this is a normal attack
+                $this->player->damage($playerDamage);
+            }
+            else {
+                // The damage done to the player is 0.
+                // We interpretate this as a miss.
+            }
+        }
+    }
+    
+    /**
+     * Returns the damage done to the player and to the monster.
+     * @return array [playerDamage, monsterDamage, playerAttack]
+     */
+    protected function calculateDamage(): array
+    {
+        $monsterDefense = $this->monster->getDefense();
+        $monsterAttack = $this->monster->getAttack();
+        $playerDefense = $this->player->getDefense();
+        $playerAttack = $this->player->getAttack();
+        
+        $monsterDamage = 0;
+        $playerDamage = 0;
+        
+        while ($monsterDamage === 0 && $playerDamage === 0) {
+            $atk = $playerAttack;
+            
+            // Critical hit probablity is derived from the old e_rand() function.
+            // e_rand(1, 3) == 3 has a probablity of ~25%.
+            if ($this->diceBag->chance(0.25)) {
```

Comment:
I thought about creating a BattleLog which contains the important messages. I'll look into your stuff.